src/hotspot/share/opto/predicates.cpp

```diff
@@ -147,3 +149,126 @@ Node* PredicateBlock::skip_regular_predicates(Node* regular_predicate_proj, Deop
   }
   return entry;
 }
+
+// This strategy clones the OpaqueLoopInit and OpaqueLoopStride nodes.
+class CloneStrategy : public TransformStrategyForOpaqueLoopNodes {
+  PhaseIdealLoop* const _phase;
+  Node* const _new_ctrl;
+
+ public:
+  CloneStrategy(PhaseIdealLoop* phase, Node* new_ctrl)
+      : _phase(phase),
+        _new_ctrl(new_ctrl) {}
+  NONCOPYABLE(CloneStrategy);
+
+  Node* transform_opaque_init(OpaqueLoopInitNode* opaque_init) const override {
+    return _phase->clone_and_register(opaque_init, _new_ctrl)->as_OpaqueLoopInit();
+  }
+
+  Node* transform_opaque_stride(OpaqueLoopStrideNode* opaque_stride) const override {
+    return _phase->clone_and_register(opaque_stride, _new_ctrl)->as_OpaqueLoopStride();
+  }
+};
+
+// Creates an identical clone of this Template Assertion Predicate Expression (i.e.cloning all nodes from the Opaque4Node
+// to and including the OpaqueLoop* nodes). The cloned nodes are rewired to reflect the same graph structure as found for
+// this Template Assertion Predicate Expression. The cloned nodes get 'new_ctrl' as ctrl. There is no other update done
+// for the cloned nodes. Return the newly cloned Opaque4Node.
+Opaque4Node* TemplateAssertionPredicateExpression::clone(Node* new_ctrl, PhaseIdealLoop* phase) {
+  CloneStrategy clone_init_and_stride_strategy(phase, new_ctrl);
+  return clone(clone_init_and_stride_strategy, new_ctrl, phase);
+}
+
+// Class to collect data nodes from a source to target nodes by following the inputs of the source node recursively.
+// The class takes a node filter to decide which input nodes to follow and a target node predicate to start backtracking
+// from. All nodes found on all paths from source->target(s) are returned in a Unique_Node_List (without duplicates).
+class DataNodesOnPathsToTargets : public StackObj {
+  typedef bool (*NodeCheck)(const Node*);
+
+  // Node filter function to decide if we should process a node or not while searching for targets.
+  NodeCheck _node_filter;
+  // Function to decide if a node is a target node (i.e. where we should start backtracking). This check should also
+  // trivially pass the _node_filter.
+  NodeCheck _is_target_node;
+  Unique_Node_List _collected_nodes; // The resulting node collection of all nodes on paths from source->target(s).
```

Comment:
```suggestion
  // The resulting node collection of all nodes on paths from source->target(s).
  Unique_Node_List _collected_nodes;
```
Just cosmetic. All other comments are on their own lines too.